fdbclient/NativeAPI.actor.cpp

```diff
@@ -2375,15 +2375,44 @@ Reference<WatchMetadata> DatabaseContext::getWatchMetadata(int64_t tenantId, Key
 }
 
 void DatabaseContext::setWatchMetadata(Reference<WatchMetadata> metadata) {
-	watchMap[std::make_pair(metadata->parameters->tenant.tenantId, metadata->parameters->key)] = metadata;
+	const WatchMapKey key(metadata->parameters->tenant.tenantId, metadata->parameters->key);
+	watchMap[key] = metadata;
+	// NOTE Here we do *NOT* update/reset the reference count for the key, see the source code in getWatchFuture
+}
+
+int32_t DatabaseContext::increaseWatchRefCount(const int64_t tenantID, KeyRef key) {
+	const WatchMapKey mapKey(tenantID, key);
+	if (watchCounterMap.count(mapKey) == 0) {
+		watchCounterMap[mapKey] = 0;
+	}
+	const auto count = ++watchCounterMap[mapKey];
+	return count;
+}
+
+int32_t DatabaseContext::decreaseWatchRefCount(const int64_t tenantID, KeyRef key) {
+	const WatchMapKey mapKey(tenantID, key);
+	if (watchCounterMap.count(mapKey) == 0) {
```

Comment:
This could do fewer lookups by using `.find` and then operating on the iterator.